### ISSUE TITLE
aggregation response arrays?

### ISSUE THREAD

**Issue by arussellsaw** [aerospike/aerospike-client-php#48]
i have an aggregation i'm running with a stream udf for filtering rows, when executed in aql i get many responses back from my udf, matching my criteria (defined in json and passed to the udf). when using the php client i get an array with a single 'bins' item, with one incorrectly matched row. is there some special behaviour required for multiple value response stream udfs? the documentation doesn't help at all really, so if anyone could point me in the right direction that would be much appreciated. 

here's a simplified version of my problem:

aql> aggregate some.stream_udf('{"criteria":[{"key":"val"},{"key":"val"}]}') on test.all where key = val

-------row--------
-------row--------
-------row--------

$where = \Aerospike::predicateEquals("key", "val");
$aerospike->aggregate("test", "all", $where, "some", "stream_udf", array('{"criteria":[{"key":"val"},{"key":"val"}]}'), $output);

var_dump($output);
// returns an array with a bins item with a single row not matching criteria

**Comment by rbotzer** [comment]
Hey Alex, can you quote your Lua, and results in both cases (AQL, and the PHP script)? That will help with reproducing and debugging the problem.

**Comment by arussellsaw** [comment]
looks like now it's now returning a single row matching the criteria (bug in my json encoding), code incoming.

**Comment by rbotzer** [comment]
Alrighty, thanks.

**Comment by arussellsaw** [comment]
```php
//this string is matching bins sequentially based on match logic and 
$limitations = '{"banner_limits":[{"limit_comparison":"==","limit_type":"geo_c_code","limit_logic":"and","limit_data":["uk"],"limit_order":0},{"limit_comparison":"=~","limit_type":"dev_limit","limit_logic":"and","limit_data":["android"],"limit_order":1}]}';
$where = \Aerospike::predicateEquals("agency_key", $agency);
$status = $this->as->aggregate("forecast", "all", $where, "limitations", "match", array($limitations), $output);
var_dump($output);
/*array(3) {
  ["bins"]=>
  array(12) {
    ["z_id"]=>
    int(33)
    ["server"]=>
    string(4) "BETA"
    ["agency_key"]=>
    string(6) "BETA15"
    ["zone_key"]=>
    string(6) "BETA33"
    ["campaign_key"]=>
    string(7) "BETA767"
    ["dev_limit"]=>
    string(7) "android"
    ["vol"]=>
    int(2)
    ["geo_c_code"]=>
    string(2) "uk"
    ["a_id"]=>
    int(15)
    ["win_c_id"]=>
    int(767)
    ["pp"]=>
    int(26)
    ["win_c_type"]=>
    string(3) "rtb"
  }
  ["key"]=>
  array(4) {
    ["ns"]=>
    string(8) "forecast"
    ["set"]=>
    string(3) "all"
    ["key"]=>
    NULL
    ["digest"]=>
    NULL
  }
  ["metadata"]=>
  NULL
}

AQL output looks like this: 

+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| match                                                                                                                                                                                                                       |
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| {"z_id":303, "server":"BETA", "agency_key":"BETA15", "zone_key":"BETA303", "campaign_key":"BETA1114", "dev_limit":"computer", "vol":69, "geo_c_code":"it", "a_id":15, "win_c_id":1114, "pp":2, "win_c_type":"rtb"}      |
| {"z_id":146, "server":"BETA", "agency_key":"BETA15", "zone_key":"BETA146", "campaign_key":"BETA1355", "dev_limit":"computer", "vol":0, "geo_c_code":"pa", "a_id":15, "win_c_id":1355, "pp":12, "win_c_type":"rtb"}      |
| {"z_id":55, "server":"BETA", "agency_key":"BETA15", "zone_key":"BETA55", "campaign_key":"BETA750", "dev_limit":"androidtablet", "vol":0, "geo_c_code":"uk", "a_id":15, "win_c_id":750, "pp":38, "win_c_type":"rtb"}     |
... goes on like this for several thousand lines.
*/
```

**Comment by arussellsaw** [comment]
https://gist.github.com/arussellsaw/e2a44b3010156ac5d2fc for the lua

**Comment by rbotzer** [comment]
That's definitely a bug, and I doubt it'll be hard to reproduce. I'll open an internal bug ticket for it later today, and raise its priority. T

**Comment by rbotzer** [comment]
On this now. The internal ticket, for your reference, is AER-3948. I'm also going to pull the results up out of _bins_, and get rid of _key_, _metadata_, _bins_, because they're meaningless in this context and out of sync with the other clients.

**Comment by rbotzer** [comment]
This has been implemented in [release 3.4.0](https://github.com/aerospike/aerospike-client-php/releases/tag/3.4.0). Thanks again, Alex.

```php
Applying a stream UDF to records with ages between 0 and 200 using aggregate()≻ [✓] 
Counts of characters in their first two centuries, grouped by workplace:
array(1) {
  [0]=>
  array(2) {
    ["HAL Institute"]=>
    int(1)
    ["Planet Express"]=>
    int(4)
  }
}
   111	$where = $db->predicateBetween("age", 0, 200);
   112	$status = $db->aggregate("test", "characters", $where, "example_aggregate_udf", "group_count", array("workplace"), $result);
   113	if($status !== Aerospike::OK) {
   114	    echo standard_fail($db);
   115	} else {
   116	    echo success();
   117	    echo "Counts of characters in their first two centuries, grouped by workplace:\n";
   118	    var_dump($result);
   119	}

Showing characters WHERE ages between 20 and 200 AND workplace='Planet Express' ≻ [✓] 
The characters expressed as name => age pairs:
array(4) {
  [0]=>
  array(1) {
    ["Dr. Amy Wong"]=>
    int(34)
  }
  [1]=>
  array(1) {
    ["Prof. Hubert J Farnsworth"]=>
    int(173)
  }
  [2]=>
  array(1) {
    ["Tarunga Leela"]=>
    int(39)
  }
  [3]=>
  array(1) {
    ["Philip J Fry"]=>
    int(40)
  }
}
   125	$where = $db->predicateBetween("age", 20, 200);
   126	$status = $db->aggregate("test", "characters", $where, "example_aggregate_udf", "ages", array('workplace', 'Planet Express'), $result);
   127	if($status !== Aerospike::OK) {
   128	    echo standard_fail($db);
   129	} else {
   130	    echo success();
   131	    echo "The characters expressed as name => age pairs:\n";
   132	    var_dump($result);
   133	}
```
